README.md

[![Arduino CI](https://github.com/RobTillaart/PCA9553/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/PCA9553/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/PCA9553/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/PCA9553/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/PCA9553/actions/workflows/jsoncheck.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/PCA9553/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/PCA9553.svg?maxAge=3600)](https://github.com/RobTillaart/PCA9553/releases)


# PCA9553

Arduino library for PCA9553 I2C 8 bit PWM LED driver, 4 channel.


## Description

This experimental library is to control the I2C PCA9553 PWM extender.
This device has two possible hardcoded I2C addresses 0x62 and 0x63, 
see table below.
If you need to connect more PCA9553 devices to a single I2C bus you 
need a I2C multiplexer like https://github.com/RobTillaart/TCA9548.

The device has two PWM "generators", 0 and 1, and one can set the 
duty cycle and the frequency by means of a pre-scaler. 

Every output channel 0..3 can select to which PWM generator it is 
connected, or if it is set to ON or OFF.

The output channels can also be used as generic GPIO, however that
is not implemented in the first release.


#### From datasheet

Maximum output sink current is 25 mA per bit and 100 mA per package.

Power-On Reset (POR) initializes the registers to their default state, 
all zeroes, causing the bits to be set HIGH (LED off).


#### I2C adresses

|  type        |  Address  |
|:-------------|:---------:|
|  PCA9553/01  |  0x62     | 
|  PCA9553/02  |  0x63     |


#### Related

- https://github.com/RobTillaart/PCA9551  (8 channel)
- https://github.com/RobTillaart/PCA9552  (16 channel)
- https://github.com/RobTillaart/PCA9553  (4 channel)

Follow up series
- https://github.com/RobTillaart/PCA9634 (8 channel)
- https://github.com/RobTillaart/PCA9635 (16 channel)
- https://github.com/RobTillaart/PCA9685_RT (16 channel)



## Interface

```cpp
#include "PCA9553.h"
```


#### Constructor

- **PCA9553(uint8_t deviceAddress, TwoWire \*wire = &Wire)** Constructor with I2C device address,  Address = 0x62 or 0x63.
and optional the Wire interface as parameter.
- **bool begin()** initializes the library after startup.
Returns true if device address is available on I2C bus.
- **bool begin(int sda, int scl)**
idem, ESP32 ESP8266 only.
- **bool isConnected()** checks if address is available on I2C bus.
- **uint8_t getAddress()** returns I2C address.
- **uint8_t outputCount()** returns the number of channels = 4.
- **uint8_t reset()**


#### GPIO

- **uint8_t getInput()** read all current output levels.
- **void pinMode(uint8_t led, uint8_t mode)** set LED pin to INPUT or OUTPUT.
- **void digitalWrite(uint8_t led, uint8_t val)** set LED pin HIGH or LOW.
- **uint8_t digitalRead(uint8_t led)** read current state of LED pin.


#### Prescaler Frequency

Get and set the pre-scaler of the PWM generator.

- **void setPrescaler(uint8_t gen, uint8_t psc = 255)** set pre-scaler for generator, default 255.
- **uint8_t getPrescaler(uint8_t gen)** get the set value.

gen = 0 or 1

The main oscillator frequency can be divided by a pre-scaler.
The period of ```BLINK = (PSC + 1) / 44```
This gives the output a blink range of 0.172 Hz to 44 Hz.

Some "magic" pre-scalers.  (to be confirmed).

|  psc  |  Period  |  Frequency  |
|:-----:|:--------:|:-----------:|
|    0  |  0.0227  |  44.00 Hz   |
|    1  |  0.0455  |  22.00 Hz   |
|    3  |  0.0909  |  11.00 Hz   |
|    7  |  0.1818  |   5.50 Hz   |
|   10  |  0.250   |   4.00 Hz   |
|   21  |  0.500   |   2.00 Hz   |
|   43  |  1.000   |   1.00 Hz   |
|   87  |  2.000   |   0.50 Hz   |
|  175  |  4.000   |   0.25 Hz   |
|  219  |  5.000   |   0.20 Hz   |
|  255  |  5.818   |  0.172 Hz   |


#### PWM

Get and set the duty cycle of the PWM generator.

- **void setPWM(uint8_t gen, uint8_t psc = 128)** set PWM for generator, 
default value = 128.
- **uint8_t getPWM(uint8_t gen)** get the set value.

gen = 0 or 1

The duty cycle of ```BLINK = (256 - PWM) / 256```

|  pwm  |  Duty Cycle  |
|:-----:|:------------:|
|    0  |     0%       |
|   64  |    25%       |
|  128  |    50%       |
|  192  |    75%       |
|  255  |   100%       |

Note: one might need a Gamma brightness correction - https://github.com/RobTillaart/GAMMA


#### Output Mode

- **uint8_t setOutputMode(uint8_t pin, uint8_t mode)** set the mode for 
the selected output pin to one of 4 modi operandi.
See table below.
  - pin == 0..3, mode == 0..3, see table below.
  - returns 0 if OK
  - returns error code if parameter is out of range.
- **uint8_t getOutputMode(uint8_t led)** returns current setting.
  - pin == 0..3
  - returns mode, see table below.
  - returns error code if parameter is out of range. 

|  Define             |  Value  |  Output pin          |
|:--------------------|:-------:|:---------------------|
|  PCA9553_MODE_LOW   |    0    |  is set LOW (LED on)
|  PCA9553_MODE_HIGH  |    1    |  is set high-impedance (LED off; default)
|  PCA9553_MODE_PWM0  |    2    |  blinks at PWM0 rate
|  PCA9553_MODE_PWM1  |    3    |  blinks at PWM1 rate


#### Power On Reset

The PCA9553 will keep its settings as long as it is powered on. 
This means it can start with an previous configuration when uploading 
two different sketches short after each other.

To handle this the library has a **reset()** function which sets
the device in the Power On state.


#### Error codes

These are kept similar to PCA9635 et al error codes.

|  Error code             |  Value  |  Description           |
|:------------------------|:-------:|:-----------------------|
|  PCA9553_OK             |   0x00  |  Everything went well
|  PCA9553_ERROR          |   0xFF  |  Generic error
|  PCA9553_ERR_WRITE      |   0xFE  |
|  PCA9553_ERR_CHAN       |   0xFD  |  output pin out of range / channel error
|  PCA9553_ERR_MODE       |   0xFC  |  mode parameter out of range.
|  PCA9553_ERR_REG        |   0xFB  |
|  PCA9553_ERR_I2C        |   0xFA  |


## Future

#### Must

- improve documentation
- test with hardware

#### Should

- improve error handling
  - return values, where etc.
- test
  - reset function


#### Could


#### Wont (on request)

- no usage of autoincrement register
- percent interface for PWM
- time interface for prescaler
- default setup in begin (what how)




CHANGELOG.md
# Change Log PCA9553

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.2.1] - 2023-09-25
- refactor API, begin();
- update readme.md;
- update examples

----

## [0.2.1] - 2023-09-25
- add Wire1 support for ESP32
- update readme.md

## [0.2.0] - 2023-07-17
- refactor interface (breaking)
  - for compatibility with PCA9552, PCA9551
- add **uint8_t getAddress()**
- add **void digitalWrite(led, value)**
- add **uint8_t digitalRead(led)**
- add **reset()**
- rename setLEDSource to **setOutputMode()**
- rename channelCount to **outputCount()**
- add #defines for modi
- add some error handling
- add examples
- update readme.md
- internal refactor

----

## [0.1.0] - 2023-07-16
- initial version




test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2023-07-16
// PURPOSE: unit tests for the PCA9553 I2C LED driver
//          https://github.com/RobTillaart/PCA9553
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// ----------------------------
// assertEqual(expected, actual);               // a == b
// assertNotEqual(unwanted, actual);            // a != b
// assertComparativeEquivalent(expected, actual);    // abs(a - b) == 0 or (!(a > b) && !(a < b))
// assertComparativeNotEquivalent(unwanted, actual); // abs(a - b) > 0  or ((a > b) || (a < b))
// assertLess(upperBound, actual);              // a < b
// assertMore(lowerBound, actual);              // a > b
// assertLessOrEqual(upperBound, actual);       // a <= b
// assertMoreOrEqual(lowerBound, actual);       // a >= b
// assertTrue(actual);
// assertFalse(actual);
// assertNull(actual);

// // special cases for floats
// assertEqualFloat(expected, actual, epsilon);    // fabs(a - b) <= epsilon
// assertNotEqualFloat(unwanted, actual, epsilon); // fabs(a - b) >= epsilon
// assertInfinity(actual);                         // isinf(a)
// assertNotInfinity(actual);                      // !isinf(a)
// assertNAN(arg);                                 // isnan(a)
// assertNotNAN(arg);                              // !isnan(a)

#include <ArduinoUnitTests.h>


#include "PCA9553.h"


unittest_setup()
{
  fprintf(stderr, "PCA9553_LIB_VERSION: %s\n", (char *) PCA9553_LIB_VERSION);
}


unittest_teardown()
{
}


unittest(test_constants)
{
  fprintf(stderr, "\nregisters");
  assertEqual(PCA9553_INPUT      , 0x00);
  assertEqual(PCA9553_PSC0       , 0x01);
  assertEqual(PCA9553_PWM0       , 0x02);
  assertEqual(PCA9553_PSC1       , 0x03);
  assertEqual(PCA9553_PWM1       , 0x04);
  assertEqual(PCA9553_LS0        , 0x05);

  fprintf(stderr, "\nerrorcodes");
  assertEqual(PCA9553_OK         , 0x00);
  assertEqual(PCA9553_ERROR      , 0xFF);
  assertEqual(PCA9553_ERR_WRITE  , 0xFE);
  assertEqual(PCA9553_ERR_CHAN   , 0xFD);
  assertEqual(PCA9553_ERR_MODE   , 0xFC);
  assertEqual(PCA9553_ERR_REG    , 0xFB);
  assertEqual(PCA9553_ERR_I2C    , 0xFA);
}


unittest(test_constructor)
{
  PCA9553 pca(0x62);

  assertEqual(4, pca.outputCount());
  assertEqual(0x62, pca.getAddress());
}

//  need mock up for more tests.
//  low prio.

unittest_main()


//  -- END OF FILE --



PCA9553.cpp
//
//    FILE: PCA9553.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2023-07-16
// VERSION: 0.2.0
// PURPOSE: Arduino library for for I2C PCA9553 4 channel PWM
//     URL: https://github.com/RobTillaart/PCA9553


#include "PCA9553.h"


//////////////////////////////////////////////////////////////
//
//  Constructor
//
PCA9553::PCA9553(const uint8_t deviceAddress, TwoWire *wire)
{
  _address = deviceAddress;
  _wire    = wire;
  _outputCount = 4;
  _error = PCA9553_OK;
}


#if defined (ESP8266) || defined(ESP32)
bool PCA9553::begin(int sda, int scl)
{
  _wire = &Wire;
  if ((sda < 255) && (scl < 255))
  {
    _wire->begin(sda, scl);
  } else {
    _wire->begin();
  }
  if (! isConnected()) return false;
  return true;
}
#endif


bool PCA9553::begin()
{
  _wire->begin();
  if (! isConnected()) return false;
  return true;
}


bool PCA9553::isConnected()
{
  _wire->beginTransmission(_address);
  _error = _wire->endTransmission();
  return (_error == 0);
}


uint8_t PCA9553::reset()
{
  //  not most efficient
  setPrescaler(0, 0);   //  44 Hz
  setPrescaler(1, 0);   //  44 Hz
  setPWM(0, 128);       //  50%
  setPWM(1, 128);       //  50%
  setOutputMode(0, 0);  //  LOW
  setOutputMode(1, 0);  //  LOW
  setOutputMode(2, 0);  //  LOW
  setOutputMode(3, 0);  //  LOW

  return PCA9553_OK;
}


uint8_t PCA9553::getAddress()
{
  return _address;
}


uint8_t PCA9553::outputCount()
{
  return _outputCount;
}


/////////////////////////////////////////////////////
//
//  GPIO
//
uint8_t PCA9553::getInput()
{
  return readReg(PCA9553_INPUT);
}


void PCA9553::pinMode(uint8_t pin, uint8_t mode)
{
  if (mode != OUTPUT) setOutputMode(pin, PCA9553_MODE_HIGH);
}


void PCA9553::digitalWrite(uint8_t pin, uint8_t val)
{
  if (val == LOW) setOutputMode(pin, PCA9553_MODE_LOW);
  else            setOutputMode(pin, PCA9553_MODE_HIGH);
}


uint8_t PCA9553::digitalRead(uint8_t pin)
{
  uint8_t value = readReg(PCA9553_INPUT);
  if ((value >> pin) & 0x01) return HIGH;
  return LOW;
}


/////////////////////////////////////////////////////
//
//  PRESCALERS
//
void PCA9553::setPrescaler(uint8_t gen, uint8_t psc)
{
  if (gen == 0) writeReg(PCA9553_PSC0, psc);
  else          writeReg(PCA9553_PSC1, psc);
}


uint8_t PCA9553::getPrescaler(uint8_t gen)
{
  if (gen == 0) return readReg(PCA9553_PSC0);
  else          return readReg(PCA9553_PSC1);
}


/////////////////////////////////////////////////////
//
//  PWM
//
void PCA9553::setPWM(uint8_t gen, uint8_t pwm)
{
  if (gen == 0) writeReg(PCA9553_PWM0, pwm);
  else          writeReg(PCA9553_PWM1, pwm);
}


uint8_t PCA9553::getPWM(uint8_t gen)
{
  if (gen == 0) return readReg(PCA9553_PWM0);
  else          return readReg(PCA9553_PWM1);
}


/////////////////////////////////////////////////////
//
//  OUTPUT MODE - MUX SELECTION
//
uint8_t PCA9553::setOutputMode(uint8_t pin, uint8_t mode)
{
  if (pin >= _outputCount)
  {
    _error = PCA9553_ERR_CHAN;
    return _error;
  }
  if (mode > 3)
  {
    _error = PCA9553_ERR_MODE;
    return _error;
  }

  uint8_t ledSelect = readReg(PCA9553_LS0);
  ledSelect &= ~(0x03 << (pin * 2));
  ledSelect |= (mode << (pin * 2));

  return writeReg(PCA9553_LS0, ledSelect);
}


uint8_t PCA9553::getOutputMode(uint8_t pin)
{
  if (pin >= _outputCount)
  {
    _error = PCA9553_ERR_CHAN;
    return _error;
  }

  uint8_t ledSelect = readReg(PCA9553_LS0);
  uint8_t mode = (ledSelect >> (pin * 2)) & 0x03;
  return mode;
}


uint8_t PCA9553::getLastError()
{
  uint8_t e = _error;
  _error = PCA9553_OK;
  return e;
}



/////////////////////////////////////////////////////
//
//  PRIVATE
//
uint8_t PCA9553::writeReg(uint8_t reg, uint8_t value)
{
  _wire->beginTransmission(_address);
  _wire->write(reg);
  _wire->write(value);
  _error = _wire->endTransmission();

  if (_error == 0) _error = PCA9553_OK;
  else _error = PCA9553_ERR_I2C;
  return _error;
}


uint8_t PCA9553::readReg(uint8_t reg)
{
  _wire->beginTransmission(_address);
  _wire->write(reg);
  _error = _wire->endTransmission();
  if (_error == 0) _error = PCA9553_OK;
  else
  {
    _error = PCA9553_ERR_I2C;
    return 0;
  }
  if (_wire->requestFrom(_address, (uint8_t)1) != 1)
  {
    _error = PCA9553_ERR_I2C;
    return 0;
  }
  _error = PCA9553_OK;
  return _wire->read();
}


//  -- END OF FILE --



PCA9553.h
#pragma once
//
//    FILE: PCA9553.h
//  AUTHOR: Rob Tillaart
//    DATE: 2023-07-16
// VERSION: 0.2.0
// PUPROSE: Arduino library for for I2C PCA9553 4 channel PWM
//     URL: https://github.com/RobTillaart/PCA9553


#include "Arduino.h"
#include "Wire.h"


#define PCA9553_LIB_VERSION         (F("0.2.0"))


//  REGISTERS
#define PCA9553_INPUT       0x00
#define PCA9553_PSC0        0x01
#define PCA9553_PWM0        0x02
#define PCA9553_PSC1        0x03
#define PCA9553_PWM1        0x04
#define PCA9553_LS0         0x05


//  MUX OUTPUT MODES
#define PCA9553_MODE_LOW      0
#define PCA9553_MODE_HIGH     1
#define PCA9553_MODE_PWM0     2
#define PCA9553_MODE_PWM1     3


//  ERROR CODES (not all used yet)
#define PCA9553_OK          0x00
#define PCA9553_ERROR       0xFF
#define PCA9553_ERR_WRITE   0xFE
#define PCA9553_ERR_CHAN    0xFD
#define PCA9553_ERR_MODE    0xFC
#define PCA9553_ERR_REG     0xFB
#define PCA9553_ERR_I2C     0xFA


/////////////////////////////////////////////////////
//
//  CLASS
//
class PCA9553
{
public:
  explicit PCA9553(const uint8_t deviceAddress, TwoWire *wire = &Wire);

#if defined (ESP8266) || defined(ESP32)
  bool     begin(int sda, int scl);
#endif
  bool     begin();
  bool     isConnected();
  uint8_t  reset();

  uint8_t  getAddress();
  uint8_t  outputCount();

  //  GPIO
  uint8_t  getInput();
  void     pinMode(uint8_t pin, uint8_t mode);
  void     digitalWrite(uint8_t pin, uint8_t value);
  uint8_t  digitalRead(uint8_t pin);

  //  PRESCALERS
  void     setPrescaler(uint8_t generator, uint8_t prescaler = 0);
  uint8_t  getPrescaler(uint8_t generator);

  //  PWM
  void     setPWM(uint8_t generator, uint8_t pwm = 128);
  uint8_t  getPWM(uint8_t generator);


  //  OUTPUT MODE - MUX SELECTION
  //  pin:   0..3
  //  mode:
  //    0 = output is set LOW (LED on)
  //    1 = output is set high-impedance (LED off; default)
  //    2 = output blinks at PWM0 rate
  //    3 = output blinks at PWM1 rate
  uint8_t  setOutputMode(uint8_t pin, uint8_t mode);
  uint8_t  getOutputMode(uint8_t pin);

  //  ERROR
  uint8_t  getLastError();


private:
  uint8_t  writeReg(uint8_t reg, uint8_t value);
  uint8_t  readReg(uint8_t reg);

  uint8_t  _address;
  uint8_t  _channelCount;
  uint8_t  _error;

  TwoWire* _wire;
};


//  -- END OF FILE --

